hngprojects/hng_boilerplate_java_web
# [FEAT] Implementation of Resource Management

### HOW TO TEST : Endpoints for Resources Management

### Description
These API endpoints allows for managing resources within the application. These endpoints will allow for retrieving resources, editing them, creating them and deleting them. 

### How to Test
Access the `GET /api/v1/resources/articles` endpoint to search for published resources returns default page limit for first page
For example:  ` http://localhost:8080/api/v1/resources/articles`
Expected result: Success.
Example response: 

```java
{
    "data": [
        {
            "id": "8cdf7502-1bde-413a-a2b8-7c2490f3a373",
            "title": "Walmart",
            "description": "Sales",
            "image": "hng_images.png",
            "date": "2024-08-14",
            "published": true
        },
        {
            "id": "21594800-85d6-41a6-ba73-32dfac984d17",
            "title": "Amazon",
            "description": "goods",
            "image": "hng_images.png",
            "date": "2024-08-14",
            "published": true
        }
    ],
    "totalPages": 10,
    "currentPage": 0``` 
``` 

Access the ` GET /api/v1/resources/search `  endpoint and encounter a server error.
Expected result: A 500 error is returned.
```java
Example response: {
  "status": "error",
  "message": "Internal Server Error. Please try again later.",
  "status_code": 500
}
``` 

Access the ` POST /api/v1/resources/create `  endpoint to create resources.
For example: `http://localhost:8080/api/v1/resource/create `
Expected result: Created Successfully.
Response:

```java
{
    "message": "How to use youtube  Created Successfully",
    "resourceData": {
        "id": "21594800-85d6-41a6-ba73-32dfac984d17",
        "title": "How to use youtube",
        "description": "Using youtube",
        "image": "hng_images.png",
        "date": "2024-08-14",
        "published": true
    }
``` 

Access the ` PATCH  /api/v1/resources/edit `  endpoint to edit already edited resource
For example: `http://localhost:8080/api/v1/resource/edit `
Expected result: Edited Succesfully.
Response:
```java
{
    "message": "How to use youtube  Edited Successfully",
    "resourceData": {
        "id": "21594800-85d6-41a6-ba73-32dfac984d17",
        "title": "How to use youtube",
        "description": "Using youtube",
        "image": "hng_images.png",
        "date": "2024-08-14",
        "published": true
    }
``` 

Access the ` DELETE /api/v1/resources/delete/{Id} `  endpoint to delete a resource that doesn’t exist
For example: `http://localhost:8080/api/v1/resource/{id}`
Expected result: Success.

```java
{
    "message": "This resource does not exist",
    "error": "Resource with 3cf9e963-ecd0-42e8-9cd3-c5b0780af8a0 not found",
    "status_code": 404
}
``` 

Access the ` GET  /api/v1/resources/{Id} `  endpoint to get resource with an id
For example: `http://localhost:8080/api/v1/resource/{id}`
Expected result: Success.

```java
    "resourceData": {
        "id": "21594800-85d6-41a6-ba73-32dfac984d17",
        "title": "How to use youtube",
        "description": "Using youtube",
        "image": "hng_images.png",
        "date": "2024-08-14",
        "published": true
       }
``` 

Access the ` PATCH  /api/v1/resources/unpublish/{Id} `  endpoint to unpublish a resource
For example: `http://localhost:8080/api/v1/resource/unpublish/{id}`
Expected result: Success.

```java
      "message": "How to use youtube unpublished Successfully"
    "resourceData": {
        "id": "21594800-85d6-41a6-ba73-32dfac984d17",
        "title": "How to use youtube",
        "description": "Using youtube",
        "image": "hng_images.png",
        "date": "2024-08-14",
        "published": false
         
  }
``` 


Access the ` PATCH  /api/v1/resources/publish/{Id} `  endpoint to publish a resource
For example: `http://localhost:8080/api/v1/resource/publish/{id}`
Expected result: Success.

```java
      "message": "How to use youtube published Successfully"
    "resourceData": {
        "id": "21594800-85d6-41a6-ba73-32dfac984d17",
        "title": "How to use youtube",
        "description": "Using youtube",
        "image": "hng_images.png",
        "date": "2024-08-14",
        "published": true
     }
``` 

Access the ` GET  /api/v1/resources/search`  endpoint for an admin to search all resources, published or not.
For example: `http://localhost:8080/api/v1/resource/publish/{id}`
Expected result: Success.

``` java
"data": [
        {
            "id": "4f4214fb-ec1f-4096-9c9e-5ecbd2a80304",
            "title": "How to make ogbono soup",
            "description": " A breakdown to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": false
        },
        {
            "id": "9420e6ee-eec5-4190-80f1-9417b2b05b05",
            "title": "How to make owoh soup",
            "description": " A guide to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": false
        },
        {
            "id": "358b9095-6aad-4770-b1c8-b388c361f837",
            "title": "How to make banga soup",
            "description": " A guide to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": false
        },
        {
            "id": "1ba89099-449e-44fa-be3c-1cc88c5d6ccf",
            "title": "How to make okra soup",
            "description": " A guide to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": true
        },
        {
            "id": "8f26b20a-9d70-4f6d-aaaa-b32512d702c7",
            "title": "How to make white soup",
            "description": " A guide to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": true
        }
``` 
Access the ` GET  /api/v1/resources/published`  endpoint for an admin to search all published resources
For example: `http://localhost:8080/api/v1/resource/publish/{id}`
Expected result: Success.

``` java
"data": [
        {
            "id": "4f4214fb-ec1f-4096-9c9e-5ecbd2a80304",
            "title": "How to make ogbono soup",
            "description": " A breakdown to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": true
        },
        {
            "id": "9420e6ee-eec5-4190-80f1-9417b2b05b05",
            "title": "How to make owoh soup",
            "description": " A guide to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": true
        }
``` 


Access the ` GET  /api/v1/resources/unpublished`  endpoint for an admin to search all published resources
For example: `http://localhost:8080/api/v1/resource/publish/{id}`
Expected result: Success.

``` java
"data": [
        {
            "id": "4f4214fb-ec1f-4096-9c9e-5ecbd2a80304",
            "title": "How to make ogbono soup",
            "description": " A breakdown to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": false
        },
        {
            "id": "9420e6ee-eec5-4190-80f1-9417b2b05b05",
            "title": "How to make owoh soup",
            "description": " A guide to making one of Nigeria's best delicacy",
            "image": "hng_images.png",
            "date": "2024-08-16",
            "published": false
        }
```
